ProjFS.Mac/PrjFSKext/KauthHandler.cpp

```diff
@@ -216,6 +241,162 @@ KEXT_STATIC void UseMainForkIfNamedStream(
     }
 }
 
+KEXT_STATIC bool InitPendingRenames()
+{
+    if (version_major >= 18) // Only need to track renames on Mojave and newer
+    {
+        s_renameLock = SpinLock_Alloc();
+        s_maxPendingRenames = 8; // Arbitrary choice, should be maximum number of expected concurrent threads performing renames, but array will resize on demand
+        s_pendingRenameCount = 0;
+        s_pendingRenames = Memory_AllocArray<PendingRenameOperation>(s_maxPendingRenames);
+        if (!SpinLock_IsValid(s_renameLock) || s_pendingRenames == nullptr)
+        {
+            return false;
+        }
+
+        Array_DefaultInit(s_pendingRenames, s_maxPendingRenames);
+    }
+    
+    return true;
+}
+
+KEXT_STATIC void CleanupPendingRenames()
+{
+    if (version_major >= 18)
+    {
+        if (SpinLock_IsValid(s_renameLock))
+        {
+            SpinLock_FreeMemory(&s_renameLock);
+        }
+
+        if (s_pendingRenames != nullptr)
+        {
+            assert(s_pendingRenameCount == 0);
+            Memory_FreeArray(s_pendingRenames, s_maxPendingRenames);
+            s_pendingRenames = nullptr;
+            s_maxPendingRenames = 0;
+        }
+    }
+}
+
+KEXT_STATIC void ResizePendingRenames(uint32_t newMaxPendingRenames)
+{
+    PendingRenameOperation* newArray = Memory_AllocArray<PendingRenameOperation>(newMaxPendingRenames);
```

Comment:
I don't see the need for using a `_noblock` fast path here because this code is going to be called so rarely anyway. And as we call this only when no lock is held, the blocking version is safe.